just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java

```diff
@@ -576,52 +580,63 @@ private MediaSource getAudioSource(final Object json) {
     }
 
     private MediaSource decodeAudioSource(final Object json) {
-        Map<?, ?> map = (Map<?, ?>)json;
-        String id = (String)map.get("id");
-        switch ((String)map.get("type")) {
-        case "progressive":
-            return new ProgressiveMediaSource.Factory(buildDataSourceFactory())
-                    .createMediaSource(new MediaItem.Builder()
-                            .setUri(Uri.parse((String)map.get("uri")))
-                            .setTag(id)
-                            .build());
-        case "dash":
-            return new DashMediaSource.Factory(buildDataSourceFactory())
-                    .createMediaSource(new MediaItem.Builder()
-                            .setUri(Uri.parse((String)map.get("uri")))
-                            .setMimeType(MimeTypes.APPLICATION_MPD)
-                            .setTag(id)
-                            .build());
-        case "hls":
-            return new HlsMediaSource.Factory(buildDataSourceFactory())
-                    .createMediaSource(new MediaItem.Builder()
-                            .setUri(Uri.parse((String)map.get("uri")))
-                            .setMimeType(MimeTypes.APPLICATION_M3U8)
-                            .build());
-        case "silence":
-            return new SilenceMediaSource.Factory()
-                    .setDurationUs(getLong(map.get("duration")))
-                    .setTag(id)
-                    .createMediaSource();
-        case "concatenating":
-            MediaSource[] mediaSources = getAudioSourcesArray(map.get("children"));
-            return new ConcatenatingMediaSource(
-                    false, // isAtomic
-                    (Boolean)map.get("useLazyPreparation"),
-                    decodeShuffleOrder(mapGet(map, "shuffleOrder")),
-                    mediaSources);
-        case "clipping":
-            Long start = getLong(map.get("start"));
-            Long end = getLong(map.get("end"));
-            return new ClippingMediaSource(getAudioSource(map.get("child")),
-                    start != null ? start : 0,
-                    end != null ? end : C.TIME_END_OF_SOURCE);
-        case "looping":
-            Integer count = (Integer)map.get("count");
-            MediaSource looperChild = getAudioSource(map.get("child"));
-            return new LoopingMediaSource(looperChild, count);
-        default:
-            throw new IllegalArgumentException("Unknown AudioSource type: " + map.get("type"));
+        Map<?, ?> map = (Map<?, ?>) json;
+        String id = (String) map.get("id");
+
+        DrmSessionManager drmSessionManager = null;
+        if (map.get("drm") != null) {
+            Map<?, ?> drm = (Map<?, ?>) map.get("drm");
+            if (drm.get("clearKey") != null) {
+                drmSessionManager = new DefaultDrmSessionManager.Builder()
```

Comment:
In future, how would this approach be extended to support a custom DrmSessionManager?

Reply:
New methos should be added to the Dart helper calls(DRMHelper) to create a Map with the required key value pairs for the DRM. The platform specific code should then check if those properties are available and instantiate the DRMManager accordingly.(Android)

Reply:
I understand the design, but a custom DrmCessionManager should be able to implement callbacks with custom logic, particularly to support use cases like offline playback.

Reply:
The DRM Clearkey is support for offline playback. The key/kid pair is provided to the DRM manager and if media asset contains the KID then corresponding key is taken.(call back happens natively)

For callbacks this is done in online mode where the URL is called to relevant server. The idea would be the specify URL(if possible) and token information in the drm map which would be provided to the DRMSesssion manager.